Add and enforce org storage quota

tw4l opened the pull request:
Fixes #1060 
Fixes #930 

This PR introduces org storage quotas to the backend and uses them in the frontend. The org frontend now polls the backend every 10 seconds to find out if the org storage quota has been reached via a new API endpoint.

In the backend, the org storage quota is now checked before starting new crawls and uploads, and they are not started if it has been reached. Current storage usage in bytes is saved in the `Organization` model, calculated on the collective size of archived item and profile files. Routines have also been added to make sure this value is kept up to date as archived items and profiles are added and deleted.

The quota can be configured in the orgs list in the frontend:

<img width="514" alt="Screen Shot 2023-08-24 at 5 19 12 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/09fecbd7-68b5-4be9-b573-10db53cf11d1">

When the quota has been configured to a value greater than 0 and it's reached, the following changes happen in the frontend:

The new workflow button is disabled, with an "Org Storage Full" tooltip:

<img width="221" alt="Screen Shot 2023-08-24 at 5 19 24 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/1af79706-6100-4d14-af2d-d6977b5f2913">

Same for the Upload WACZ button in Archived Items:

<img width="159" alt="Screen Shot 2023-08-24 at 5 19 33 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/46c0beac-5ab2-495e-89ec-78c849fc2ee2">

The option to create a new workflow in the collection editor disappears:

<img width="1107" alt="Screen Shot 2023-08-24 at 5 19 46 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/b75a403a-df5e-406c-91be-b5caf73ddec1">

Attempting to run a crawl via an action results in the following toast notification:

<img width="445" alt="Screen Shot 2023-08-24 at 5 23 17 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/e9445c04-efc4-4141-b7dd-ef70797662b2">

I've tested that setting the quota higher then resets the frontend, enabling the buttons and allowing the actions again.

@Shrinks99 I think wish the Dashboard plans to show storage usage and quota, this should be enough for the frontend for now but would like your input.

tw4l committed:
Track bytesStored in organization

- Add migration to pre-calculate based on size of crawlfiles and profilefiles
- Add methods to increase or decrease org storage when crawl or profile files
are added or deleted

tw4l committed:
Don't start new crawls if storage quota reached

tw4l committed:
Add endpoint to return boolean indicating if storage quota reached

tw4l committed:
Fix migration docstring

tw4l committed:
Add org storage quota to frontend

- Add to orgs-list quotas
- Disable buttons when storage quota is met
- Show toast notification when attempting to run a crawl when org
storage quota is met

tw4l committed:
Add response model for storage quota endpoint

SuaYoo commented on frontend/src/pages/org/index.ts:
```diff
@@ -73,6 +76,9 @@ export class Org extends LiteElement {
   orgId!: string;
 
+  @property({ type: Boolean })
+  orgStorageQuotaReached: boolean = false;
```
Since this attribute doesn't depend on a reactive property of `btrix-app`, you could make this a private property of `btrix-org`:
```suggestion
  @state()
  private orgStorageQuotaReached = false;
```

SuaYoo commented on frontend/src/index.ts:
```diff
@@ -642,6 +642,7 @@ export class App extends LiteElement {
           .params=${this.viewState.params}
           orgId=${orgId}
+          orgStorageQuotaReached="false"
           orgPath=${orgPath.split(orgId)[1]}
```
See comment in `frontend/src/pages/org/index.ts` for rationale:
```suggestion
          orgPath=${orgPath.split(orgId)[1]}
```

SuaYoo commented on frontend/src/pages/org/collection-editor.ts:
```diff
@@ -99,6 +99,9 @@ export class CollectionEditor extends LiteElement {
   orgId!: string;
 
+  @property({ type: Boolean })
+  orgStorageQuotaReached!: boolean;
```
Prefer to set a default boolean value if the app can still function (versus something like an ID string, in which there is no default value without breaking functionality:)
```suggestion
  orgStorageQuotaReached = false;
```

SuaYoo commented on frontend/src/pages/org/collections-new.ts:
```diff
@@ -33,6 +36,7 @@ export class CollectionsNew extends LiteElement {
       <btrix-collection-editor
         .authState=${this.authState}
         orgId=${this.orgId}
+        .orgStorageQuotaReached=${this.orgStorageQuotaReached!}
```
Use `?` to [prefix boolean attributes](https://lit.dev/docs/templates/expressions/#boolean-attribute-expressions):
```suggestion
        ?orgStorageQuotaReached=${this.orgStorageQuotaReached}
```

SuaYoo commented on frontend/src/pages/org/crawls-list.ts:
```diff
@@ -209,13 +212,19 @@ export class CrawlsList extends LiteElement {
-                  @click=${() => (this.isUploadingArchive = true)}
+                <sl-tooltip
+                  content=${msg("Org Storage Full")}
+                  ?disabled=${this.orgStorageQuotaReached === false}
```
It seems like this has the opposite effect of disabling the button when the storage quota is not reached.
```suggestion
                  ?disabled=${this.orgStorageQuotaReached}
```

tw4l replied:
The tooltip and the button have reverse disabled logic. We want the button to be disabled if the quota is reached, so that the user can't upload a WACZ. But we want the tooltip disabled if the quota *isn't* yet reached, so that it only displays when the quota is reached. Took me a second to get my head around.

SuaYoo replied:
Gotcha!

SuaYoo commented on frontend/src/pages/org/crawls-list.ts:
```diff
@@ -209,13 +212,19 @@ export class CrawlsList extends LiteElement {
+                  <sl-button
+                    size="small"
+                    @click=${() => (this.isUploadingArchive = true)}
+                    ?disabled=${this.orgStorageQuotaReached === true}
```
```suggestion
                    ?disabled=${this.orgStorageQuotaReached}
```

SuaYoo commented on frontend/src/pages/org/workflow-detail.ts:
```diff
@@ -1410,6 +1428,17 @@ export class WorkflowDetail extends LiteElement {
+    if (this.orgStorageQuotaReached === true) {
+      this.notify({
+        message: msg(
+          "The org has reached its storage limit. Delete any archived items that are un-needed to free up space, or contact us to purchase a plan with more storage."
```
```suggestion
          "The org has reached its storage limit. Delete any archived items that are unneeded to free up space, or contact us to purchase a plan with more storage."
```
Should we put in contact info here? "contact us at _@webrecorder.net" or "contact us from the Webrecorder contact page"?

tw4l replied:
Leaving this open. @Shrinks99 - any thoughts?

Shrinks99 replied:
Leaning towards not putting links in toasts but could go for it?  This also seems like something that will be dealt with by including storage purchase options in the app eventually.  Additionally, hardcoding our contact info makes it vendor specific for self-hosted deployments... which maybe we don't care about considering at this time, but maybe we do?

tw4l replied:
@Shrinks99 so maybe keep as-is for now, and come back to add this later if we see a need?

SuaYoo commented on frontend/src/pages/org/index.ts:
```diff
@@ -82,6 +88,17 @@ export class Org extends LiteElement {
+  // Use to cancel requests
+  private getStorageQuotaReachedController: AbortController | null = null;
+
+  private timerId?: number;
```
To prevent lit from performing updates after the component is removed from the DOM, you can clear the timer in `disconnectedCallback()`:
```
  disconnectedCallback() {
    window.clearInterval(this.timerId);
    super.disconnectedCallback();
  }
  ```

SuaYoo commented on frontend/src/pages/org/index.ts:
```diff
@@ -112,6 +129,48 @@ export class Org extends LiteElement {
           icon: "exclamation-octagon",
         });
       }
+      this.checkIfStorageQuotaReached();
```
Since `this.getOrgStorageQuotaReached` is below `await this.getOrg`, but doesn't depend on `this.getOrg` to resolve, you could move this to line 121 above the try-catch block and parallelize requests a bit.

SuaYoo reviewed (changes requested):
Great start! I added a few changes as suggestions, but figured it might be easier for you to find-and-replace multiple instances rather than add a bunch of suggestions to batch. To summarize:

1. `orgStorageQuotaReached` can be an internal state property of `Org`, since the value is not updated in `App`

2. Prefer to set reasonable default to all boolean properties--typescript will infer the type and you don't have to use the [typescript `!` postfix](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-type-assertions):
```diff
   @property({ type: Boolean })
- orgStorageQuotaReached!: boolean;
+ orgStorageQuotaReached = false;
```

3. Boolean attributes should be prefixed with `?` (also removed postfix here):
```diff
- .orgStorageQuotaReached=${this.orgStorageQuotaReached!}
+ ?orgStorageQuotaReached=${this.orgStorageQuotaReached}
```

4. Once the boolean attributes are fixed, the tooltips will be disabled since `this.orgStorageQuotaReached === false` means the quota is not reached:
```diff
- ?disabled=${this.orgStorageQuotaReached === false}
+ ?disabled=${this.orgStorageQuotaReached}
```

5. To prevent lit from performing updates after navigating away from an org view (e.g. on log out,) you will want to clear the timer with `disconnectedCallback`.

tw4l committed:
Make code review improvements

- Update migration version number to 0015
- Remove polling and API storage quota endpoint
- Return storageQuotaReached from API endpoints and use it to
set local state via events
- Add 403 exception to uploads and profiles if quota is reached
- Make crawl deletion consistent with other endpoints
- Store quota in bytes on backend and use base10 in frontend
- Remove duplicate inc_org_bytes_stored in crawl deletion

tw4l commented:
Thanks @SuaYoo - I really appreciate the help skilling up on the frontend! I've made the requested changes, minus adding contact info and changing the tooltip logic, since I think you'll find as weird as it is, it's working as intended:

> Once the boolean attributes are fixed, the tooltips will be disabled since this.orgStorageQuotaReached === false means the quota is not reached:

That's by design as we want the tooltips to only show (i.e. be enabled) when the button is disabled, because that means that the quota has been reached.

ikreymer commented on backend/btrixcloud/migrations/migration_0014_org_storage_usage.py:
```diff
@@ -0,0 +1,51 @@
+"""
+Migration 0014 - Calculate and store org storage usage
```
We already have a 0014 migration now, should rename this to 0015

ikreymer commented on backend/btrixcloud/basecrawls.py:
```diff
@@ -166,6 +168,7 @@ async def delete_crawls(
             size += await self._delete_crawl_files(crawl, org)
             if crawl.get("cid"):
                 cids_to_update.add(crawl.get("cid"))
+            await inc_org_bytes_stored(self.orgs_db, org.id, -size)
```
Should we move this to be after the deletion operation, just in case, and consistent with other deletion updates?

tw4l replied:
Yeah good call! That way it only runs once too

ikreymer commented on backend/btrixcloud/db.py:
```diff
@@ -15,7 +15,7 @@
 
 
-CURR_DB_VERSION = "0013"
+CURR_DB_VERSION = "0014"
```
now onto 0015!

```suggestion
CURR_DB_VERSION = "0015"
```

ikreymer commented on backend/btrixcloud/models.py:
```diff
@@ -609,6 +609,14 @@ class OrgQuotas(BaseModel):
 
     maxConcurrentCrawls: Optional[int] = 0
     maxPagesPerCrawl: Optional[int] = 0
+    storageQuotaGB: Optional[int] = 0
```
Should we make this be base2 or base10 GiB? or just store bytes and decide later/need more flexibility?

I guess storing bytes might require special handling on frontend to convert back.

tw4l replied:
Took it to Discord to discuss whether we should use base2 or base10 consistently. If the former, I think storing in GB (technically GiB) is a bit simpler as it matches what the frontend can set.

tw4l replied:
Decided on base10, will make changes

ikreymer commented on backend/btrixcloud/orgs.py:
```diff
@@ -402,6 +457,15 @@ async def update_quotas(
+        tags=["organizations"],
+        response_model=OrgStorageQuotaReachedOut,
+    )
+    async def get_storage_quota_reached(org: Organization = Depends(org_dep)):
```
The separate endpoint is necessary for polling of the quota and easier access on the frontend, right?
Otherwise, seems like it could be part of the standard org GET api also.

tw4l replied:
True! The polling may in fact be unnecessary...

ikreymer commented:
Nice work! Looks good and just a couple comments (see above).

One other question I have is if frontend polling makes sense for this, since it's something that the state shouldn't be changing too often, unless a user is close to quota and then ends up deleting things when they reach it, but ideally that's not something that's a common enough use case? (and then we can send an event to reload the quota/org state after archived item or profile delete?) Or maybe something to consider for future iteration, since this is already in place, but just want to be mindful of the background polling that we're doing.

tw4l commented:
> One other question I have is if frontend polling makes sense for this, since it's something that the state shouldn't be changing too often, unless a user is close to quota and then ends up deleting things when they reach it, but ideally that's not something that's a common enough use case? (and then we can send an event to reload the quota/org state after archived item or profile delete?) Or maybe something to consider for future iteration, since this is already in place, but just want to be mindful of the background polling that we're doing.

Hm, that's a good point, this is a lot of polling for something that should change infrequently. Perhaps it was more useful for me in testing than it would be in production.

Not sure we wanna add receiving events just for this, but in the scenario you mention a browser refresh/cache clearing would also resolve the issue. I'll look into modifying it to just use the value from the GET org API endpoint and see how that goes.

tw4l commented:
Ah, now I remember why I had the polling there - it's the other way around. If the quota is exceeded, without polling (or a webhook, as you mention, but that would require us adding a system for receiving webhook events too), if a user doesn't refresh their browser or clear their cache, they could continue to try to create crawls from the frontend. The backend would stop them but the buttons wouldn't be disabled in the frontend, which might cause confusion as it'd result in an opaque error.

Up to you but I'm tempted to leave as-is unless we see it become a problem.

ikreymer commented on backend/btrixcloud/crawlconfigs.py:
```diff
@@ -681,6 +689,9 @@ async def run_now(self, cid: str, org: Organization, user: User):
             )
 
+        if await self.org_ops.storage_quota_reached(org):
+            raise HTTPException(status_code=507, detail="storage_quota_reached")
```
Hm, actually had to look up 507 error code, seems like a 4xx code is more appropriate here? 402?
since its a user quota? or maybe just 403?
Some discussion here:
https://softwareengineering.stackexchange.com/questions/288376/recommended-http-status-code-for-plan-limit-exceeded-response

tw4l replied:
403 seems reasonable? and we can check the detail in the frontend?

tw4l replied:
Went with 403 and `storage_quota_reached` detail

ikreymer commented:
> Ah, now I remember why I had the polling there - it's the other way around. If the quota is exceeded, without polling (or a webhook, as you mention, but that would require us adding a system for receiving webhook events too), if a user doesn't refresh their browser or clear their cache, they could continue to try to create crawls from the frontend. The backend would stop them but the buttons wouldn't be disabled in the frontend, which might cause confusion as it'd result in an opaque error.
> 
> Up to you but I'm tempted to leave as-is unless we see it become a problem.

We should we return an error with `storage_quota_reached` (and a designated status code, TBD) for all operations that use incur storage:
- starting new crawl
- starting an upload
- starting a new browser profile

Then, the api handler can just check for that and send something like:
 `this.dispatchEvent(new CustomEvent("update-storage-quota", { detail: {exceeded: true} }))` which will update the top-level storageQuota and have it propagate down (or do an api fetch, though we just need a bool if exceeded or not), so may not even need to do an extra fetch.
Similarly, after:
- deleting an archived item
- deleting a profile
Can also send an event to trigger a call to the api to refresh the quota. Think that will be a bit more robust overall?

ikreymer commented on backend/btrixcloud/profiles.py:
```diff
@@ -176,14 +183,16 @@ async def commit_to_profile(
             {"_id": profile.id}, {"$set": profile.to_dict()}, upsert=True
         )
 
+        await inc_org_bytes_stored(self.orgs, oid, file_size)
```
We should error out of `commit_to_profile()` right away if the storage quota is reached.

ikreymer commented:
What if on the backend we had:
- any api that would result in `inc_org_bytes_storage` being called with positive value (crawl, upload, profile saving) may raise `storage_quota_exceeded` and error out quickly if storage quota exceeds before it starts.
- any api that calls `inc_org_bytes_storage` also returns a `storageQuotaReached` bool reflecting if storage quota is exceeded after it has run.

This way, an upload can complete if it starts below quota, but if it reaches quota afterwards, can be reflected right away

SuaYoo commented:
> We should we return an error with `storage_quota_reached` (and a designated status code, TBD) for all operations that use incur storage

+1 for storage quota error code

tw4l commented:
> What if on the backend we had:
> 
>     * any api that would result in `inc_org_bytes_storage` being called with positive value (crawl, upload, profile saving) may raise `storage_quota_exceeded` and error out quickly if storage quota exceeds before it starts.

I'm not opposed but this is different from what we'd previously said, that we would let crawls that would exceed the quota go but stop afterwards. Open to change it though!

tw4l commented:
Converted back to draft while I work on the changes

tw4l commented:
@ikreymer @SuaYoo ready for re-review! I've done a lot of manual testing locally but wouldn't mind a few more eyes on it too :) Thanks!

tw4l committed:
Merge branch 'main' into issue-1060-org-storage-quota

tw4l committed:
Linting

SuaYoo reviewed (commented):
Nice one! I think dispatching `storage_quota_reached` is a common enough side effect that you could handle it in `apiFetch()` in `LiteElement.ts`, similar to how the `need-login` event is dispatched on line 128. You could check that `options?.method` exists, and is not `GET`, so that the event is dispatched on all POST, PUT, PATCH and DELETEs.

ikreymer commented on backend/btrixcloud/orgs.py:
```diff
@@ -253,6 +253,15 @@ async def get_max_pages_per_crawl(self, org: Organization):
         """Return org-specific max pages per crawl setting or 0."""
         return await get_max_pages_per_crawl(self.orgs, org.id)
 
+    async def inc_bytes_stored(self, org: Organization, size: int):
```
Is there a reason to have this call to the non-class function?

Previously, added non-class functions so can use in operator, which didn't have the ops classes, but now that they do, should probably just keep things in the class?

ikreymer replied:
oh, i see, i guess the OrgOps isn't yet available everywhere that its called, can do a cleanup pass on this another time, so feel free to disregard for now.

ikreymer reviewed (approved):
backend changes look good!

tw4l committed:
Move storage-quota-update event into LiteElement's apiFetch

tw4l commented:
> Nice one! I think dispatching `storage_quota_reached` is a common enough side effect that you could handle it in `apiFetch()` in `LiteElement.ts`, similar to how the `need-login` event is dispatched on line 128. You could check that `options?.method` exists, and is not `GET`, so that the event is dispatched on all POST, PUT, PATCH and DELETEs.

Good call! I had to keep the other check in `file-uploader` because it's using XHR rather than apiFetch() but otherwise this greatly simplified it! I've tested locally and it's still working well. Also verified that the update is limited only to the org in question.

tw4l committed:
Remove storage quota from collections

SuaYoo commented:
> I had to keep the other check in file-uploader because it's using XHR rather than apiFetch()

@tw4l In `main` I see that it was also kept in browser-profiles-detail, browser-profiles-new, workflow-detail and workflows-list, any specific reason for those?

tw4l commented:
@ikreymer those are responding to a specific error and inform the message that's shown in the toast sent by that component so I thought it appropriate to keep them there, but we could refactor that as a follow-up.

tw4l commented:
Ah, could handle the dispatching of the event at least on a 403 - I'll change that in a follow-up PR.
